conf/pvers.c

```diff
@@ -805,19 +811,27 @@ static int conf_pver_virtual_create(const struct m0_fid *fid,
 }
 
 /**
- * Check if failures at any level has reached max allowed failures.
+ * Check if failures at any level has reached or exceeded max allowed failures.
  */
-static bool has_reached_max_failure(struct m0_conf_pver *pv,
-				    const uint32_t      *srecd)
+static int tolerance_failure_cmp(struct m0_conf_pver *pv,
+				 const uint32_t      *srecd)
 {
-	int i = 0;
+	int      i = 0;
+	int      result = MAX_FAILURES_NOT_REACHED;
+	uint32_t tolerance ;
 
 	while(i < M0_CONF_PVER_HEIGHT) {
-		if (srecd[i] == pv->pv_u.subtree.pvs_tolerance[i])
-			return true;
+		tolerance = pv->pv_u.subtree.pvs_tolerance[i];
+		/* Ignore the case of srecd[i] == tolerance == 0. */
+		if (srecd[i] > 0 && srecd[i] == tolerance)
+			result = MAX_FAILURES_REACHED;
+		else if (srecd[i] > tolerance) {
```

Comment:
I am assuming that:
if (tolerance == 0)
   return MAX_FAILURES_NOT_REACHED
else {
  /* proceed with srecd[] based checks */
}
If this assumption is correct then the else-if block can be entered even when tolerance == 0 which is not correct.

Reply:
[4/1/2022 6:07 PM] Abhishek Saha
Hi, I did not get your commenthttps://github.com/Seagate/cortx-motr/pull/1571#discussion_r840433364

[4/1/2022 6:11 PM] Abhishek Saha
if the question is about why we are we going to else if block if tolerance == 0, is to cover the case when eg: node tolerance is set to 0, but we still get a node failure, so we want to return that MAX_FAILURE_EXCEEDED as we have more failure than we can tolerate

[4/1/2022 6:12 PM] Mehul Joshi
hmm, ok .. I thought that if the tolerance is 0 then it means we have not defined any MAX failures for that component.
[4/1/2022 6:14 PM] Mehul Joshi
for example, what will happend if we have N+K as 4+2 but out tolerance vector is [ 0 0 0 0 0 ] ?

[4/1/2022 6:15 PM] Mehul Joshi
will one disk failure cross MAX failures?

[4/1/2022 6:15 PM] Abhishek Saha
yes

[4/1/2022 6:15 PM] Mehul Joshi
then instead of degraded  mode we will say damaged, right?

[4/1/2022 6:21 PM] Abhishek Saha
yes, for disk level we stripe the object to have only one unit at a time, so losing K disks is inherently tolerated by motr, but it is not the same for controller or enclosure level. depending upon the configuration. If we don't set any tolerances the striping for eg: controller could end with more than K units of a parity group in it. So unless tolerance is explicitly set, we treat any failures as unsupported.

[4/2/2022 8:24 AM] Mehul Joshi
OK. So, how do I setup my N, K and tolerance vector to say something like - "I don't care how many nodes/racks/sites go offline as long as I don't have more than K disks offline" ?

[10:07 AM] Abhishek Saha
AFAIK, there is a way to ignore failures at any level. If we get failures they are treated as real failures and motr acts accordingly.

[10:10 AM] Mehul Joshi
I don't understand .. could you please explain with reference to my question above?

[10:16 AM] Abhishek Saha
I don't care how many nodes/racks/sites go offline as long as I don't have more than K disks offlineThere is no way to not care, about for eg: node failures, If node tolerance is set to 0 and we get a node failure. Motr will go into degraded mode and then degraded IO will fail because we exceeded the tolerance. The recent changes I made, just better reflects this in bytecount.

[10:18 AM] Mehul Joshi
OK.